clang/lib/Sema/Sema.cpp

```diff
@@ -1931,6 +1931,17 @@ void Sema::checkTypeSupport(QualType Ty, SourceLocation Loc, ValueDecl *D) {
   if (isUnevaluatedContext() || Ty.isNull())
     return;
 
+  // The original idea behind checkTypeSupport function is that unused
+  // declarations can be replaced with an array of bytes of the same size during
+  // codegen, however this is not yet implemented for SYCL and cause production
+  // of incorrect SPIR-V.
+  // In addition such replacement doesn't seem to be possible for types without
+  // constant byte size like zero length arrays. So, do a deep check for SYCL.
```

Comment:
What about flexible array members?

Reply:
Can be diagnosed this way as well. I thought about that, but I left it for next patches to keep this one simpler/smaller.